chips/stm32f4xx/src/chip_specific.rs

```diff
@@ -0,0 +1,166 @@
+        pub const PLL_MIN_FREQ_MHZ: usize = __PLL_MIN_FREQ_MHZ;
+    }
+
+    #[cfg(any(feature = "stm32f412"))]
+    const __APB1_FREQUENCY_LIMIT_MHZ: usize = 50;
+    #[cfg(any(feature = "stm32f429", feature = "stm32f446"))]
+    #[cfg(not(feature = "cargo-clippy"))]
+    const __APB1_FREQUENCY_LIMIT_MHZ: usize = 45;
+    #[cfg(any(feature = "stm32f401"))]
+    #[cfg(not(feature = "cargo-clippy"))]
+    const __APB1_FREQUENCY_LIMIT_MHZ: usize = 42;
+
+    /// Maximum allowed APB1 frequency in MHz
+    pub const APB1_FREQUENCY_LIMIT_MHZ: usize = __APB1_FREQUENCY_LIMIT_MHZ;
+
+    /// Maximum allowed APB2 frequency in MHz
+    // APB2 frequency limit is twice the APB1 frequency limit
+    pub const APB2_FREQUENCY_LIMIT_MHZ: usize = APB1_FREQUENCY_LIMIT_MHZ << 1;
+
+    #[cfg(any(feature = "stm32f412"))]
+    const __SYS_CLOCK_FREQUENCY_LIMIT_MHZ: usize = 100;
+    #[cfg(any(feature = "stm32f429", feature = "stm32f446"))]
+    #[cfg(not(feature = "cargo-clippy"))]
+    const __SYS_CLOCK_FREQUENCY_LIMIT_MHZ: usize = 168;
+    #[cfg(any(feature = "stm32f401"))]
+    #[cfg(not(feature = "cargo-clippy"))]
+    const __SYS_CLOCK_FREQUENCY_LIMIT_MHZ: usize = 84;
+
+    /// Maximum allowed system clock frequency in MHz
+    pub const SYS_CLOCK_FREQUENCY_LIMIT_MHZ: usize = __SYS_CLOCK_FREQUENCY_LIMIT_MHZ;
+}
+
+/// Chip-specific flash code
+pub mod flash_specific {
+    #[cfg(any(
+        feature = "stm32f401",
+        feature = "stm32f412",
+        feature = "stm32f429",
+        feature = "stm32f446"
+    ))]
+    #[derive(Copy, Clone, PartialEq, Debug)]
+    /// Enum representing all the possible values for the flash latency
+    pub(crate) enum FlashLatency {
+        /// 0 wait cycles
+        Latency0,
+        /// 1 wait cycle
+        Latency1,
+        /// 2 wait cycles
+        Latency2,
+        /// 3 wait cycles
+        Latency3,
+        /// 4 wait cycles
+        Latency4,
+        /// 5 wait cycles
+        Latency5,
+        /// 6 wait cycles
+        Latency6,
+        /// 7 wait cycles
+        Latency7,
+        /// 8 wait cycles
+        Latency8,
+        /// 9 wait cycles
+        Latency9,
+        /// 10 wait cycles
+        Latency10,
+        /// 11 wait cycles
+        Latency11,
+        /// 12 wait cycles
+        Latency12,
+        /// 13 wait cycles
+        Latency13,
+        /// 14 wait cycles
+        Latency14,
+        /// 15 wait cycles
+        Latency15,
+    }
+
+    // The number of wait cycles depends on two factors: system clock frequency and the supply
+    // voltage. Currently, this method assumes 2.7-3.6V voltage supply (default value).
+    // TODO: Take into the account the power supply
+    //
+    // The number of wait states varies from chip to chip.
+    pub(crate) fn get_number_wait_cycles_based_on_frequency(frequency_mhz: usize) -> FlashLatency {
+        // feature = "stm32f401"
+        // feature = "stm32f429"
+        // feature = "stm32f446"
+        #[cfg(not(feature = "stm32f412"))]
```

Comment:
I think that's the issue with the `#[cfg`-based approach as-opposed to the `cfg!(`-based approach.

From https://github.com/tock/tock/pull/3528#issuecomment-1691034340, do you prefer option (1) [which I believe is currently what this code attached to this PR does] or option (2) [which is linked from [this comment](https://github.com/tock/tock/pull/3528#issuecomment-1684269209)]. I think (2) is better and would alleviate this problem.

Reply:
For (2), do we need the ifs? Can we just have a default thing that happens which is the panic? Sort of like how the arch crate has a bunch of empty functions which only show up if the correct function can't be included because we're compiling for a host machine.

Reply:
Are you talking about functions like this one:
```rust
#[cfg(not(any(target_arch = "arm", target_os = "none")))]
pub unsafe extern "C" fn initialize_ram_jump_to_main() {
    unimplemented!()
}
```

The problem with this approach is that code compiles. The code will panic only if the `unimplemented!()` macro is called at runtime. Not sure if this is the desirable behaviour.